tests/mocha/workspace_test.js

```diff
@@ -0,0 +1,357 @@
+      this.workspace.newBlock('');
+      this.workspace.newBlock('');
+      this.workspace.clear();
+      chai.assert.equal(this.workspace.getTopBlocks(false).length, 0);
+    });
+  });
+
+  suite('getAllBlocks', function() {
+    test('Empty workspace', function() {
+      chai.assert.equal(this.workspace.getAllBlocks(true).length, 0);
+    });
+
+    test('Flat workspace one block', function() {
+      this.workspace.newBlock('');
+      chai.assert.equal(this.workspace.getAllBlocks(true).length, 1);
+    });
+
+    test('Flat workspace one block after dispose', function() {
+      var blockA = this.workspace.newBlock('');
+      this.workspace.newBlock('');
+      blockA.dispose();
+      chai.assert.equal(this.workspace.getAllBlocks(true).length, 1);
+    });
+
+    test('Flat workspace two blocks', function() {
+      this.workspace.newBlock('');
+      this.workspace.newBlock('');
+      chai.assert.equal(this.workspace.getAllBlocks(true).length, 2);
+    });
+
+    test('Clear', function() {
+      this.workspace.clear();
+      chai.assert.equal(this.workspace.getAllBlocks(true).length, 0,
+          'Clear empty workspace');
+      this.workspace.newBlock('');
+      this.workspace.newBlock('');
+      this.workspace.clear();
+      chai.assert.equal(this.workspace.getAllBlocks(true).length, 0);
+    });
+  });
+
+  suite('remainingCapacity', function() {
+    setup(function() {
+      this.workspace.newBlock('');
+      this.workspace.newBlock('');
+    });
+
+    test('No block limit', function() {
+      chai.assert.equal(this.workspace.remainingCapacity(), Infinity);
+    });
+
+    test('Under block limit', function() {
+      this.workspace.options.maxBlocks = 3;
+      chai.assert.equal(this.workspace.remainingCapacity(), 1);
+      this.workspace.options.maxBlocks = 4;
+      chai.assert.equal(this.workspace.remainingCapacity(), 2);
+    });
+
+    test('At block limit', function() {
+      this.workspace.options.maxBlocks = 2;
+      chai.assert.equal(this.workspace.remainingCapacity(), 0);
+    });
+
+    test('At block limit of 0 after clear', function() {
+      this.workspace.options.maxBlocks = 0;
+      this.workspace.clear();
+      chai.assert.equal(this.workspace.remainingCapacity(), 0);
+    });
+
+    test('Over block limit', function() {
+      this.workspace.options.maxBlocks = 1;
+      chai.assert.equal(this.workspace.remainingCapacity(), -1);
+    });
+
+    test('Over block limit of 0', function() {
+      this.workspace.options.maxBlocks = 0;
+      chai.assert.equal(this.workspace.remainingCapacity(), -2);
+    });
+  });
+
+  suite('remainingCapacityOfType', function() {
+    setup(function() {
+      this.workspace.newBlock('get_var_block');
+      this.workspace.newBlock('get_var_block');
+      this.workspace.options.maxInstances = {};
+    });
+
+    test('No instance limit', function() {
+      chai.assert.equal(this.workspace.remainingCapacityOfType('get_var_block'),
+          Infinity);
```

Comment:
Just curious what this Infinity is?

Reply:
It's a built-in global object in Javascript.